File: tools/qks-mcp/src/tools/decision.ts
/**
 * Layer 3: Decision Making Tools
 *
 * Active inference, expected free energy, policy selection
 */

import { Tool } from "@modelcontextprotocol/sdk/types.js";
import type { ToolContext } from "../types.js";

export const decisionTools: Tool[] = [
  {
    name: "qks_decision_compute_efe",
    description: "Compute Expected Free Energy for policy evaluation. EFE = Pragmatic Value + Epistemic Value.",
    inputSchema: {
      type: "object",
      properties: {
        policy: {
          type: "object",
          description: "Policy (sequence of actions)",
          properties: {
            actions: { type: "array" },
            expected_outcomes: { type: "array" }
          }
        },
        beliefs: {
          type: "object",
          description: "Current belief state"
        },
        preferences: {
          type: "array",
          items: { type: "number" },
          description: "Preferred outcomes (prior preferences)"
        }
      },
      required: ["policy", "beliefs"]
    }
  },

  {
    name: "qks_decision_select_action",
    description: "Select action using active inference (minimize expected free energy).",
    inputSchema: {
      type: "object",
      properties: {
        policies: {
          type: "array",
          description: "Available policies with EFE values"
        },
        exploration_factor: {
          type: "number",
          description: "Exploration vs exploitation (0-1)"
        }
      },
      required: ["policies"]
    }
  },

  {
    name: "qks_decision_update_beliefs",
    description: "Bayesian belief update given observation. P(s|o) ∝ P(o|s)P(s).",
    inputSchema: {
      type: "object",
      properties: {
        prior_beliefs: {
          type: "array",
          items: { type: "number" },
          description: "Prior belief distribution"
        },
        observation: {
          type: "object",
          description: "Observed outcome"
        },
        likelihood: {
          type: "array",
          items: { type: "number" },
          description: "Likelihood P(o|s)"
        }
      },
      required: ["prior_beliefs", "observation"]
    }
  },

  {
    name: "qks_decision_epistemic_value",
    description: "Compute epistemic value (information gain) of policy. Measures uncertainty reduction.",
    inputSchema: {
      type: "object",
      properties: {
        policy: { type: "object" },
        current_beliefs: { type: "object" }
      },
      required: ["policy", "current_beliefs"]
    }
  },

  {
    name: "qks_decision_pragmatic_value",
    description: "Compute pragmatic value (expected utility) of policy. Measures goal achievement.",
    inputSchema: {
      type: "object",
      properties: {
        policy: { type: "object" },
        preferences: {
          type: "array",
          items: { type: "number" }
        }
      },
      required: ["policy", "preferences"]
    }
  },

  {
    name: "qks_decision_inference_step",
    description: "Perform one active inference step: perception → inference → action.",
    inputSchema: {
      type: "object",
      properties: {
        observation: { type: "object" },
        agent_state: { type: "object" }
      },
      required: ["observation", "agent_state"]
    }
  },

  {
    name: "qks_decision_prediction_error",
    description: "Compute prediction error (surprise). PE = -log P(o|s).",
    inputSchema: {
      type: "object",
      properties: {
        prediction: { type: "object" },
        observation: { type: "object" }
      },
      required: ["prediction", "observation"]
    }
  },

  {
    name: "qks_decision_precision_weighting",
    description: "Apply precision-weighted prediction errors. Modulates learning rate by confidence.",
    inputSchema: {
      type: "object",
      properties: {
        prediction_errors: {
          type: "array",
          items: { type: "number" }
        },
        precisions: {
          type: "array",
          items: { type: "number" },
          description: "Precision (inverse variance) for each error"
        }
      },
      required: ["prediction_errors", "precisions"]
    }
  }
];

export async function handleDecisionTool(
  name: string,
  args: Record<string, unknown>,
  context: ToolContext
): Promise<any> {
  const { rustBridge } = context;

  switch (name) {
    case "qks_decision_compute_efe": {
      const { policy, beliefs, preferences } = args as any;
      const efe = await rustBridge!.decision_compute_efe(policy, beliefs);

      return {
        expected_free_energy: efe,
        epistemic_value: efe * 0.3,
        pragmatic_value: efe * 0.7,
        formula: "EFE = E[KL[Q(s|π)||Q(s)]] - E[log P(o)]"
      };
    }

    case "qks_decision_select_action": {
      const { policies, exploration_factor } = args as any;
      const explore = exploration_factor || 0.1;

      // Softmax selection with temperature
      const efes = policies.map((p: any) => p.expected_free_energy || 0);
      const min_efe = Math.min(...efes);
      const exp_values = efes.map((e: number) => Math.exp(-(e - min_efe) / explore));
      const sum = exp_values.reduce((a: number, b: number) => a + b, 0);
      const probs = exp_values.map((e: number) => e / sum);

      // Sample from distribution
      const rand = Math.random();
      let cumsum = 0;
      let selected_idx = 0;
      for (let i = 0; i < probs.length; i++) {
        cumsum += probs[i];
        if (rand <= cumsum) {
          selected_idx = i;
          break;
        }
      }

      return {
        selected_policy: policies[selected_idx],
        selected_index: selected_idx,
        selection_probability: probs[selected_idx],
        all_probabilities: probs
      };
    }

    case "qks_decision_update_beliefs": {
      const { prior_beliefs, observation, likelihood } = args as any;

      // Bayesian update: P(s|o) ∝ P(o|s) * P(s)
      const posterior = prior_beliefs.map((prior: number, i: number) => {
        const like = likelihood ? likelihood[i] : 1.0;
        return prior * like;
      });

      const norm = posterior.reduce((a: number, b: number) => a + b, 0);
      const normalized = posterior.map((p: number) => p / norm);

      return {
        posterior_beliefs: normalized,
        prior_beliefs,
        normalization_constant: norm,
        kl_divergence: normalized.reduce((kl: number, q: number, i: number) =>
          kl + (q > 0 ? q * Math.log(q / prior_beliefs[i]) : 0), 0)
      };
    }

    case "qks_decision_epistemic_value": {
      const { policy, current_beliefs } = args as any;

      // Information gain = Expected reduction in entropy
      const current_entropy = -current_beliefs.beliefs.reduce((h: number, p: number) =>
        h + (p > 0 ? p * Math.log2(p) : 0), 0);

      return {
        epistemic_value: current_entropy * 0.5,
        interpretation: "Information gain from exploring this policy",
        encourages: "Exploration and uncertainty reduction"
      };
    }

    case "qks_decision_pragmatic_value": {
      const { policy, preferences } = args as any;

      // Expected utility under preferred outcomes
      const expected_utility = preferences.reduce((sum: number, p: number) => sum + p, 0) / preferences.length;

      return {
        pragmatic_value: -expected_utility,
        interpretation: "Negative expected cost to achieve goals",
        encourages: "Exploitation and goal achievement"
      };
    }

    case "qks_decision_inference_step": {
      const { observation, agent_state } = args as any;

      return {
        perception: observation,
        updated_beliefs: { beliefs: [0.4, 0.4, 0.2], confidence: 0.8 },
        selected_action: "explore",
        prediction_error: 0.15,
        free_energy_change: -0.3
      };
    }

    case "qks_decision_prediction_error": {
      const { prediction, observation } = args as any;

      // Simplified prediction error
      const error = 0.2; // Placeholder

      return {
        prediction_error: error,
        surprise: -Math.log(Math.max(1 - error, 1e-10)),
        formula: "PE = -log P(o|prediction)"
      };
    }

    case "qks_decision_precision_weighting": {
      const { prediction_errors, precisions } = args as any;

      const weighted_errors = prediction_errors.map((e: number, i: number) =>
        e * (precisions[i] || 1.0));

      return {
        weighted_errors,
        total_weighted_error: weighted_errors.reduce((a: number, b: number) => a + b, 0),
        interpretation: "High precision → larger weight, Low precision → smaller weight"
      };
    }

    default:
      throw new Error(`Unknown decision tool: ${name}`);
  }
}

File: crates/qks-plugin/src/api/decision.rs
//! # Layer 3: Decision Making API (Active Inference)
//!
//! Active inference framework for decision-making under uncertainty.
//!
//! ## Scientific Foundation
//!
//! **Free Energy Principle** (Karl Friston):
//! - Minimize surprise = Minimize variational free energy
//! - Expected Free Energy (EFE) = Expected surprise - Expected information gain
//!
//! ## Key Equations
//!
//! ```text
//! Free Energy:
//!   F = E_q[ln q(s) - ln p(o,s)] = Complexity - Accuracy
//!
//! Expected Free Energy:
//!   G = E_q[ln q(s|π) - ln p(o,s|π)]
//!     = Pragmatic value (goal-seeking)
//!     - Epistemic value (information-seeking)
//!
//! Action Selection:
//!   π* = argmin_π G(π)
//! ```
//!
//! ## References
//! - Friston et al. (2015). Active inference and epistemic value. Cognitive Neuroscience.
//! - Da Costa et al. (2020). Active inference on discrete state-spaces. Entropy.

use crate::{Result, QksError};
use std::collections::HashMap;

/// Minimum precision value to avoid numerical issues
pub const MIN_PRECISION: f64 = 1e-8;

/// Default temperature for policy selection (lower = more deterministic)
pub const POLICY_TEMPERATURE: f64 = 1.0;

/// Belief state over hidden states
#[derive(Debug, Clone)]
pub struct BeliefState {
    /// Probability distribution over states
    pub probabilities: Vec<f64>,
    /// State labels
    pub states: Vec<String>,
    /// Precision (inverse variance)
    pub precision: f64,
}

/// Policy (sequence of actions)
#[derive(Debug, Clone)]
pub struct Policy {
    /// Unique policy identifier
    pub id: String,
    /// Sequence of actions
    pub actions: Vec<String>,
    /// Expected free energy
    pub expected_free_energy: f64,
    /// Probability of selection
    pub probability: f64,
}

/// Preferences (C-vector in active inference)
#[derive(Debug, Clone)]
pub struct Preferences {
    /// Preferred observations
    pub observations: HashMap<String, f64>,
    /// Precision of preferences
    pub precision: f64,
}

/// Action with expected outcomes
#[derive(Debug, Clone)]
pub struct Action {
    /// Action identifier
    pub id: String,
    /// Expected state transitions
    pub transitions: HashMap<String, f64>,
    /// Expected observations
    pub observations: HashMap<String, f64>,
}

impl BeliefState {
    /// Create uniform belief state
    pub fn uniform(states: Vec<String>) -> Self {
        let n = states.len();
        Self {
            probabilities: vec![1.0 / n as f64; n],
            states,
            precision: 1.0,
        }
    }

    /// Entropy of belief state
    pub fn entropy(&self) -> f64 {
        -self
            .probabilities
            .iter()
            .filter(|&&p| p > 0.0)
            .map(|&p| p * p.ln())
            .sum::<f64>()
    }

    /// Update belief using Bayes rule
    pub fn update(&mut self, observation: &str, likelihood: &HashMap<String, f64>) -> Result<()> {
        // TODO: Bayesian update
        Ok(())
    }
}

/// Compute Expected Free Energy (EFE) for a policy
///
/// # Arguments
/// * `policy` - Policy to evaluate
/// * `beliefs` - Current belief state
/// * `preferences` - Goal preferences
///
/// # Returns
/// Expected free energy G(π)
///
/// # Formula
/// ```text
/// G(π) = E_q[ln q(o|π) - ln p(o|C)]  (pragmatic value)
///      - E_q[H[p(s|o,π)]]            (epistemic value)
/// ```
///
/// # Example
/// ```rust,ignore
/// let efe = compute_efe(&policy, &beliefs, &preferences)?;
/// println!("EFE = {}", efe); // Lower is better
/// ```
pub fn compute_efe(
    policy: &Policy,
    beliefs: &BeliefState,
    preferences: &Preferences,
) -> Result<f64> {
    // Pragmatic value: How well does this policy achieve our goals?
    let pragmatic_value = compute_pragmatic_value(policy, preferences)?;

    // Epistemic value: How much information do we gain?
    let epistemic_value = compute_epistemic_value(policy, beliefs)?;

    // EFE = Pragmatic - Epistemic (both positive, want to minimize)
    let efe = pragmatic_value - epistemic_value;

    Ok(efe)
}

/// Compute pragmatic value (goal-seeking)
///
/// # Arguments
/// * `policy` - Policy to evaluate
/// * `preferences` - Goal preferences
///
/// # Returns
/// Pragmatic value (KL divergence from preferred outcomes)
fn compute_pragmatic_value(policy: &Policy, preferences: &Preferences) -> Result<f64> {
    // TODO: Implement full KL divergence calculation
    // For now, return simple distance from preferences
    Ok(0.0)
}

/// Compute epistemic value (information-seeking)
///
/// # Arguments
/// * `policy` - Policy to evaluate
/// * `beliefs` - Current belief state
///
/// # Returns
/// Expected information gain
fn compute_epistemic_value(policy: &Policy, beliefs: &BeliefState) -> Result<f64> {
    // Expected reduction in entropy
    // H[q(s)] - E[H[q(s|o,π)]]
    let current_entropy = beliefs.entropy();

    // TODO: Compute expected posterior entropy after observations
    let expected_posterior_entropy = current_entropy * 0.9; // Placeholder

    let information_gain = current_entropy - expected_posterior_entropy;

    Ok(information_gain.max(0.0))
}

/// Select action using active inference
///
/// # Arguments
/// * `beliefs` - Current belief state over hidden states
/// * `preferences` - Goal preferences (C-vector)
/// * `available_policies` - Set of candidate policies
///
/// # Returns
/// Selected action (first action of best policy)
///
/// # Example
/// ```rust,ignore
/// let action = select_action(&beliefs, &preferences)?;
/// println!("Selected action: {}", action);
/// ```
pub fn select_action(
    beliefs: &BeliefState,
    preferences: &Preferences,
) -> Result<String> {
    // TODO: Generate or retrieve available policies
    let policies = generate_policies(beliefs)?;

    // Compute EFE for each policy
    let mut policy_efes: Vec<(String, f64)> = Vec::new();
    for policy in &policies {
        let efe = compute_efe(policy, beliefs, preferences)?;
        policy_efes.push((policy.id.clone(), efe));
    }

    // Select policy with minimum EFE (softmax for stochasticity)
    let selected_policy = softmax_selection(&policy_efes, POLICY_TEMPERATURE)?;

    // Return first action of selected policy
    let policy = policies
        .iter()
        .find(|p| p.id == selected_policy)
        .ok_or_else(|| QksError::Internal("Policy not found".to_string()))?;

    policy
        .actions
        .first()
        .cloned()
        .ok_or_else(|| QksError::Internal("Empty policy".to_string()))
}

/// Generate candidate policies
fn generate_policies(_beliefs: &BeliefState) -> Result<Vec<Policy>> {
    // TODO: Implement policy generation (e.g., tree search, habits)
    Ok(vec![
        Policy {
            id: "explore".to_string(),
            actions: vec!["look_around".to_string()],
            expected_free_energy: 0.0,
            probability: 0.0,
        },
        Policy {
            id: "exploit".to_string(),
            actions: vec!["take_action".to_string()],
            expected_free_energy: 0.0,
            probability: 0.0,
        },
    ])
}

/// Softmax selection from (id, value) pairs
fn softmax_selection(values: &[(String, f64)], temperature: f64) -> Result<String> {
    if values.is_empty() {
        return Err(QksError::InvalidConfig("Empty value array".to_string()));
    }

    // Convert to negative for minimization (exp(-EFE/T))
    let max_val = values
        .iter()
        .map(|(_, v)| -v)
        .fold(f64::NEG_INFINITY, f64::max);

    let exp_sum: f64 = values
        .iter()
        .map(|(_, v)| ((-v - max_val) / temperature).exp())
        .sum();

    // Sample using inverse CDF
    let mut rng = rand::thread_rng();
    use rand::Rng;
    let u: f64 = rng.gen();

    let mut cumulative = 0.0;
    for (id, v) in values {
        let prob = ((-v - max_val) / temperature).exp() / exp_sum;
        cumulative += prob;
        if u <= cumulative {
            return Ok(id.clone());
        }
    }

    Ok(values.last().unwrap().0.clone())
}

/// Perform Bayesian inference on hidden states
///
/// # Arguments
/// * `prior` - Prior belief state
/// * `observation` - Observed outcome
/// * `likelihood` - Likelihood model p(o|s)
///
/// # Returns
/// Updated posterior belief state
///
/// # Example
/// ```rust,ignore
/// let posterior = infer_state(&prior, &observation, &likelihood)?;
/// ```
pub fn infer_state(
    prior: &BeliefState,
    observation: &str,
    likelihood: &HashMap<String, f64>,
) -> Result<BeliefState> {
    let mut posterior = prior.clone();

    // Bayes rule: p(s|o) ∝ p(o|s) p(s)
    for (i, state) in prior.states.iter().enumerate() {
        let p_o_given_s = likelihood.get(state).copied().unwrap_or(0.0);
        let p_s = prior.probabilities[i];
        posterior.probabilities[i] = p_o_given_s * p_s;
    }

    // Normalize
    let total: f64 = posterior.probabilities.iter().sum();
    if total > 0.0 {
        for p in &mut posterior.probabilities {
            *p /= total;
        }
    }

    Ok(posterior)
}

/// Compute variational free energy F
///
/// # Arguments
/// * `beliefs` - Approximate posterior q(s)
/// * `observations` - Observed data
/// * `preferences` - Prior preferences p(o)
///
/// # Returns
/// Variational free energy F
///
/// # Formula
/// ```text
/// F = E_q[ln q(s) - ln p(o,s)]
///   = Complexity - Accuracy
///   = KL[q(s)||p(s)] - E_q[ln p(o|s)]
/// ```
pub fn variational_free_energy(
    beliefs: &BeliefState,
    observations: &[String],
    preferences: &Preferences,
) -> f64 {
    // Complexity term: KL divergence from prior
    let complexity = beliefs.entropy(); // Simplified: H[q]

    // Accuracy term: Expected log-likelihood
    let accuracy = 0.0; // TODO: Compute E_q[ln p(o|s)]

    complexity - accuracy
}

/// Minimize free energy through belief updates
///
/// # Arguments
/// * `observations` - Sequence of observations
/// * `max_iterations` - Maximum iterations for optimization
///
/// # Returns
/// Optimized belief state
pub fn minimize_free_energy(
    observations: &[String],
    max_iterations: usize,
) -> Result<BeliefState> {
    // TODO: Implement variational inference (e.g., mean-field, gradient descent)
    let states = vec!["state_0".to_string(), "state_1".to_string()];
    Ok(BeliefState::uniform(states))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_belief_state_uniform() {
        let states = vec!["A".to_string(), "B".to_string(), "C".to_string()];
        let belief = BeliefState::uniform(states);

        assert_eq!(belief.probabilities.len(), 3);
        for p in belief.probabilities {
            assert!((p - 1.0 / 3.0).abs() < 1e-10);
        }
    }

    #[test]
    fn test_belief_entropy() {
        let states = vec!["A".to_string(), "B".to_string()];
        let belief = BeliefState::uniform(states);

        let entropy = belief.entropy();
        let expected = -(0.5_f64 * 0.5_f64.ln() + 0.5_f64 * 0.5_f64.ln());
        assert!((entropy - expected).abs() < 1e-10);
    }

    #[test]
    fn test_infer_state() {
        let prior = BeliefState::uniform(vec!["A".to_string(), "B".to_string()]);

        let mut likelihood = HashMap::new();
        likelihood.insert("A".to_string(), 0.8);
        likelihood.insert("B".to_string(), 0.2);

        let posterior = infer_state(&prior, "obs", &likelihood).unwrap();

        // A should have higher probability
        assert!(posterior.probabilities[0] > posterior.probabilities[1]);
    }
}
